### PR TITLE
[Website] Cache docs in memory, speed up page loads during development

### PR DESCRIPTION
**Pull request by hramos**
This only affects the website when it is hosted locally using express. The current version of the website is annoyingly sluggish, as the whole docs structure is parsed on each request.

In this PR, we store the result of extracting the Markdown sources in memory, significantly speeding up page loads. We also delay the extraction of docs until a request is made that would require them (e.g. anything that hits `/react-native/docs/*`). 

There is still a 8 second delay when the docs are first visited, as expected. This can be improved in a later PR.

## Caveats

Any changes to the docs structure may require a server restart to take effect. This is rare enough that I don't think it is a blocker. This PR significantly speeds up first page load times on the homepage and any non-docs site, and speeds up subsequent page loads on Docs. This will make for a better web development experience.

## Timing

Extracting the docs on each request takes around 8 seconds. Storing these in memory allows us to virtually remove this delay entirely. Example from a single data point (repeated tests showed these numbers to be consistent):

```
# Prior to this PR
extract docs, on every request: 8302.748ms

# With caching
extract docs, first request: 8107.119ms
extract docs, loaded from cache: 0.135ms
```

The 8 second delay is due to the following steps inside `extractDocs.js`:

```
Rendering components to Markdown: 6655.583ms
Rendering apis to Markdown: 1959.475ms
Generating styles: 212.591ms
```

### CODE REVIEW

File: website/server/extractDocs.js [hunk 1]
```diff
@@ -505,12 +505,15 @@ const styleDocs = docsList.stylesForEmbed.reduce(function(docs, filepath) {
 
 function extractDocs() {
   componentCount = 0;
+  var components = docsList.components.map(renderComponent);
```
Comment (hramos) [inline website/server/extractDocs.js]: I split these out while I was looking into the timing of each of these steps, as it allowed me to use `console.time()` and `console.timeEnd()` before/after each of these to identify the bottlenecks in extractDocs. I missed undoing this change after I removed the timing calls.

Splitting these out is not required for the main goal of this PR, if you feel this change should be reverted let me know.

**Comment by ericvicenti** [comment]
Mostly looking good. But I'm not sure why we need `memory-cache` in comparison to a normal variable

**Comment by hramos** [comment]
A normal variable would not persist across server requests.

**Comment by ericnakagawa** [comment]
LGTM!

**Pull request closed without merging**